File: 973.k-closest-points-to-origin.cpp
/*
 * @lc app=leetcode id=973 lang=cpp
 *
 * [973] K Closest Points to Origin
 */

// @lc code=start
class Solution
{
public:
    vector<vector<int>> kClosest(vector<vector<int>> &points, int k)
    {
        // * find kth smallest element from an unsorted list
        // * divide and conquer: QuickSelect()
        // make the list
        vector<pair<long long, int>> nums; // {long long = dist Square, idx}
        for (int i = 0; i < points.size(); ++i)
        {
            long long dist = points[i][0] * points[i][0] + points[i][1] * points[i][1];
            nums.push_back({dist, i});
        }

        // find kth smallest dist from origin
        long long kthSmallest = QuickSelect(0, nums.size() - 1, k, nums);
        // cout << kthSmallest << endl;

        vector<vector<int>> res;
        for (auto &pos : nums)
        {
            if (pos.first <= kthSmallest)
                res.push_back(points[pos.second]);
        }

        return res;
    }

    long long QuickSelect(int begin, int end, int k, vector<pair<long long, int>> &nums)
    {
        // 1. randomly select a pivot p
        // 2. rank(p) : split nums into 2 part(smaller, greater)
        // 3. recursively do the QuickSelect on one half
        long long pivot = nums[(begin + end) / 2].first;

        // rank (3ptr method)
        int i = begin; // it the first element after Smaller set
        int t = begin; // the element we now comp to pivot
        int j = end;   // first element before Greater set

        while (t <= j)
        {
            if (nums[t].first < pivot)
            {
                swap(nums[t], nums[i]);
                i++;
                t++;
            }
            else if (nums[t].first > pivot)
            {
                swap(nums[t], nums[j]);
                j--;
            }
            else
            {
                t++;
            }
        } //while

        int numSmaller = i - begin;

        if (k <= numSmaller)
        {
            return QuickSelect(begin, i - 1, k, nums);
        }
        else if (k <= j - begin + 1)
        {
            return pivot;
        }
        else
        {
            return QuickSelect(j + 1, end, k - (j - begin + 1), nums);
        }
    }
};
// @lc code=end

// S S S S p L L L L L

// S   S S S p p p L L L L L L
// beg       i              end
//               t
//               j

File: Sorting/Quick Select/973. K Closest Points to Origin/973. K Closest Points to Origin.java
class Solution {
    class Point{
        long dist;
        int x;
        int y;
    }
    private List<Point> ptList = new ArrayList<>();
    
    public int[][] kClosest(int[][] points, int k) 
    {   
        for(var p: points)
        {
            var pt = new Point();
            pt.x = p[0];
            pt.y = p[1];
            pt.dist = pt.x*pt.x+pt.y*pt.y;
            ptList.add(pt);
        }
        
        long kthClosest = quickSelect(0, ptList.size()-1, k);
        
        List<Point> res = new ArrayList<>();
        for(var pt: ptList)
        {
            if(pt.dist <= kthClosest)
                res.add(pt);
        }
        
        int[][] ans = new int[res.size()][];
        for(int i  = 0; i < res.size(); ++i)
        {    
            ans[i] = new int[]{res.get(i).x, res.get(i).y};
        }
        return ans;
    }
    
    private long quickSelect(int begin, int end, int k)
    {
        long pivot = ptList.get(begin).dist;
        int i = begin, j = end, t = begin;
        while(t <= j)
        {
            if(ptList.get(t).dist < pivot)
            {
                swap(i, t);
                ++t;
                ++i;
            }
            else if(ptList.get(t).dist > pivot)
            {
                swap(t, j);
                --j;
            }
            else
            {
                ++t;
            }
        }
        
        if(i-begin >= k)
            return quickSelect(begin, i-1, k);
        else if(j-begin+1 >= k)
            return pivot;
        else
            return quickSelect(j+1, end, k-(j-begin+1));
    }
    
    private void swap(int i, int j)
    {
        var temp = ptList.get(i);
        ptList.set(i, ptList.get(j));
        ptList.set(j, temp);
    }
}

// SSSS PPPPP LLLL 
// b    i   j    e




